File: src/binary-search/cpp/bsearch.cpp
/**
 * Script para apresentar um exemplo de utilização da função de busca binária em um vetor de inteiros
 */

#include "search.h"
#include <iostream>

/**
 * Função auxiliar para facilitar a exibição dos resultados
 */
void printSearchResult(int element, int result) {
    cout << "O elemento " << element << " está na posição: " << result << endl;
}

int main() {
    // Os elementos necessariamente precisam estar ordenados
    vector<int> elements = {
        0, 10, 20, 30, 40, 54, 65, 79, 123, 212, 1234
    };
    int elementToSearchOne = 65;
    int elementToSearchTwo = 887; 

    printSearchResult(elementToSearchOne, binarySearch(elements, elementToSearchOne));
    printSearchResult(elementToSearchTwo, binarySearch(elements, elementToSearchTwo));
}

File: src/binary-search/lua/bsearch.lua
--[[
    Exemplo de algoritmo de busca binária (Iterativa) implementado em Lua.
]]

function binarySearch( list, elementToSearch )
    halfIndex = 0
    startIndex = 1
    endIndex = #list

    while (startIndex <= endIndex) do
        halfIndex = (startIndex + endIndex) // 2

        if (list[halfIndex] == elementToSearch) then
            return halfIndex
        end

        if (list[halfIndex] > elementToSearch) then
            endIndex = halfIndex - 1
        else
            startIndex = halfIndex + 1
        end
    end
    return -1;
end

elements = {
    5, 6, 7, 8, 9, 10, 11, 12
}

print('Buscando um valor existente')
print(binarySearch(elements, 12))

print('Buscando um valor não existente')
print(binarySearch(elements, 45))
